### PR TITLE
feat: [CDS-82611] Traffic Routing for SMI provider BG Step [delegate]

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
Adding logic for Traffic Routing for BG step.
Refactored existing logic by extracting common logic into abstract class and adding Factory
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56787)

### CODE REVIEW

**Comment by alebuh** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreator.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+@AllArgsConstructor
+public abstract class TrafficRoutingResourceCreator {
+  private static final int K8S_RESOURCE_NAME_MAX = 253;
+  private static final String STABLE_PLACE_HOLDER = "stable";
+  private static final String STAGE_PLACE_HOLDER = "stage";
+  private static final String CANARY_PLACE_HOLDER = "canary";
+
+  protected final K8sTrafficRoutingConfig k8sTrafficRoutingConfig;
+
+  public List<KubernetesResource> createTrafficRoutingResources(String namespace, String releaseName,
```
Comment (markoharness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreator.java:41]: This is limited to two services (hosts/destinations)
We would need to be able to support more then 2.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import io.harness.annotations.dev.HarnessModuleComponent;
+import io.harness.annotations.dev.ProductModule;
+import io.harness.delegate.task.k8s.trafficrouting.K8sTrafficRoutingConfig;
+import io.harness.delegate.task.k8s.trafficrouting.ProviderType;
+
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+@CodePulse(module = ProductModule.CDS, unitCoverageRequired = true, components = {HarnessModuleComponent.CDS_K8S})
+public class TrafficRoutingResourceCreatorFactory {
```
Comment (markoharness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java:20]: This is fine, i am just thinking for TrafficRouting inherit option we do not have a provider configured. I guess we should get the provider based on the kind of resources created in one of the previous steps.

**Review by markoharness — DISMISSED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java (hunk 2 above)
Comment (alebuh) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java:20]: We can add that one easily when we start work on that logic. Just before the switch, we add a condition that if providerConfig is null -> fetch exiting resources and update them

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreator.java (hunk 1 above)
Comment (alebuh) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreator.java:41]: it's not limited, those would be used just for BG and Canary (for stable and stage/canary services). For all other cases, those value should be null

**Review by markoharness — APPROVED** [review]

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger ut0

**Comment by alebuh** [comment]
trigger uts

**Review by markoharness — DISMISSED** [review]

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger ut0, ut6, ut9

**Comment by alebuh** [comment]
trigger ut0

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/IstioTrafficRoutingResourceCreator.java [hunk 3]
```diff
@@ -96,6 +93,9 @@ private VirtualServiceSpec getVirtualServiceSpec(
 
   private List<VirtualServiceDetails> getHttpRouteSpec(
       List<TrafficRoute> routes, List<TrafficRoutingDestination> destinations, String stableName, String stageName) {
+    if (routes == null) {
+      throw new InvalidArgumentsException("Traffic Routes must be provided for Istio provider");
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/IstioTrafficRoutingResourceCreator.java:97]: nit: can we provide a more detailed error on how customer can fix this issue

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/SMITrafficRoutingResourceCreator.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                   .weight(dest.getWeight())
+                   .build())
+        .collect(Collectors.toList());
+  }
+
+  private String getRootService(SMIProviderConfig k8sTrafficRoutingConfig, String stableName) {
+    String rootService = k8sTrafficRoutingConfig.getRootService();
+    if (isEmpty(rootService)) {
+      if (isEmpty(stableName)) {
+        throw new InvalidArgumentsException("Root service should be provided in the SMI provider config");
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/SMITrafficRoutingResourceCreator.java:117]: nit: can we provide a more detailed error on how customer can fix this issue

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  public TrafficRoutingResourceCreator getResourceCreator(K8sTrafficRoutingConfig k8sTrafficRoutingConfig) {
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java:21]: nit: you could name this just create

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    ProviderType providerType = k8sTrafficRoutingConfig.getProviderConfig().getProviderType();
+    switch (providerType) {
+      case SMI:
+        return new SMITrafficRoutingResourceCreator(k8sTrafficRoutingConfig);
+      case ISTIO:
+        return new IstioTrafficRoutingResourceCreator(k8sTrafficRoutingConfig);
+      default:
+        throw new UnsupportedOperationException("Unsupported Traffic Routing provider type");
```
Comment (abosii-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java:29]: nit: add which provider type

**Review by abosii-harness — APPROVED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/IstioTrafficRoutingResourceCreator.java (hunk 3 above)
Comment (alebuh) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/IstioTrafficRoutingResourceCreator.java]: done, added HintException

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/SMITrafficRoutingResourceCreator.java (hunk 4 above)
Comment (alebuh) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/SMITrafficRoutingResourceCreator.java]: done, added HintException

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java (hunk 5 above)
Comment (alebuh) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java]: done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java (hunk 6 above)
Comment (alebuh) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/trafficrouting/TrafficRoutingResourceCreatorFactory.java]: done

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger ut0,ut6,ut7,ut8,ut9

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger ut4,ut5,ut6,ut7,ut8,ut9

**Review by abosii-harness — APPROVED** [review]

**Comment by alebuh** [comment]
trigger ut5,ut6,ut7,ut8,ut9

**Comment by alebuh** [comment]
trigger ut5, ut6, ut7, ut8, ut9

**Review by markoharness — APPROVED** [review]

**Comment by alebuh** [comment]
trigger ut5, ut6, ut7, ut8, ut9

**Comment by alebuh** [comment]
trigger ut9,review

**Comment by alebuh** [comment]
trigger ut9

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger checkstylecodeformat, ut9

**Comment by alebuh** [comment]
trigger ut9

**Comment by alebuh** [comment]
trigger ut9

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger ut7,ut9

**Pull request merged**